design/batching.md

```diff
@@ -0,0 +1,944 @@
+                    data_type: Utf8,
+                    is_nullable: false,
+                    metadata: { "rerun.kind": "entity_path" }
+                },
+            ),
+            is_nullable: false,
+            metadata: {},
+        },
+        Field {
+            name: "timepoint",
+            data_type: List(
+                Field {
+                    name: "item",
+                    data_type: Struct([
+                        Field { name: "timeline", data_type: Utf8, is_nullable: false, metadata: {} },
+                        Field { name: "type", data_type: UInt8, is_nullable: false, metadata: {} },
+                        Field { name: "time", data_type: Int64, is_nullable: false, metadata: {} },
+                    ]),
+                    is_nullable: true,
+                    metadata: {},
+                },
+            ),
+            is_nullable: false,
+            metadata: {},
+        },
+        Field {
+            name: "components",
+            data_type: Struct([
+                Field {
+                    name: "text_entry",
+                    data_type: List(
+                        Field {
+                            name: "item",
+                            data_type: Struct([
+                                Field { name: "body", data_type: Utf8, is_nullable: false, metadata: {} },
+                                Field { name: "level", data_type: Utf8, is_nullable: true, metadata: {} },
+                            ]),
+                            is_nullable: true,
+                            metadata: {
+                                "rerun.kind": "component".
+                                "rerun.component": "rerun.text_entry",
+                            },
+                        },
+                    ),
+                    is_nullable: false,
+                    metadata: {},
+                },
+            ]),
+            is_nullable: false,
+            metadata: {},
+        },
+    ],
+    metadata: {
+        "rerun.batch_id": "<BatchId>", 
+    },
+}
+```
+
+The one major difference here is that `event_id`, `entity_path` and `num_instances` join `timepoints` in having dedicated, top-level columns.
+This is important as, like timepoints, these will have to be handled separately (and deserialized into native types!) in order to drive some of the logic in the store.
+
+Lastly, we inject the `BatchId` as metadata.
+`BatchId` isn't used for any logic yet, but comes in very handy for debug purposes.
+
+At this point we might want to sort the batch by `(event_id, entity_path)`, which will greatly improve data locality once it sits in the store (see storage section below).
+
+That's also an opportunity to pre-compact the data: if two rows share the same timepoints with different components, we could potentially merge them together... that's a bit more controversial though as it means either dropping some `EventId`s, or supporting multiple `EventId`s for a single event.
```

Comment:
I think only makes sense when a single high-level SDK log call (e.g. `log_points(…, positions, colors, radii)` produces multiple components. In that case I think they should share the same `EventId` (it is the same log event).

One such-high-level log-call should produce a single `DataRow` which would become a single row in `DataTable` and `ArrowMsg`

Reply:
I actually think this might be something we want to expose more explicitly in the high level SDK in the future, for when the user wants to make it clear that a group of data came at the same time.

Perhaps something like:
```python
with rr.single_event():
    rr.log_segmentation_image("segmentation/motion", motion)
    rr.log_segmentation_image("segmentation/semantic", semantic)
```